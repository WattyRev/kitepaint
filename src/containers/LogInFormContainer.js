import React from "react";
import LogInForm from "../components/LogInForm";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { LOG_IN } from "../redux/actions";

export class LogInFormContainer extends React.Component {
  static propTypes = {
    /**
     * An identifieng string used to differentiate this from other instances
     */
    id: PropTypes.string.isRequired,
    /**
     * A function called when the log in form is submitted.
     * Provided by redux.
     */
    onSubmit: PropTypes.func.isRequired,
    /**
     * A function called when the link to the register form is clicked.
     */
    onRegister: PropTypes.func.isRequired,
    /**
     * A function called when the link to the reset password form is clicked.
     */
    onResetPassword: PropTypes.func.isRequired
  };
  state = {
    /**
     * The user-entered username value.
     * @type {String}
     */
    username: "",
    /**
     * The displayed error message.
     * @type {String}
     */
    errorMessage: null,
    /**
     * The user-entered password value.
     * @type {String}
     */
    password: "",
    /**
     * Is the register request pending?
     * @type {Boolean}
     */
    pendingRequest: false
  };

  /**
   * Handles changes to the username field by storing it in state
   * @param  {String} username The new username
   */
  handleUsernameChange = username => {
    this.setState({
      username
    });
  };

  /**
   * Handles changes to the password field by storing it in state
   * @param  {String} password The new password
   */
  handlePasswordChange = password => {
    this.setState({
      password
    });
  };

  /**
   * Handles the log in form submission
   */
  handleSubmit = () => {
    this.setState({
      pendingRequest: true
    });
    return this.props
      .onSubmit(this.state.username, this.state.password)
      .then(() => {
        this.setState({
          pendingRequest: false
        });
      })
      .catch(error => {
        this.setState({
          pendingRequest: false,
          errorMessage: error
        });
      });
  };

  render() {
    return (
      <LogInForm
        id={this.props.id}
        errorMessage={this.state.errorMessage}
        username={this.state.username}
        password={this.state.password}
        isDisabled={this.state.pendingRequest}
        onUsernameChange={this.handleUsernameChange}
        onPasswordChange={this.handlePasswordChange}
        onSubmit={this.handleSubmit}
        onRegister={this.props.onRegister}
        onResetPassword={this.props.onResetPassword}
      />
    );
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  onSubmit: LOG_IN
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LogInFormContainer);
