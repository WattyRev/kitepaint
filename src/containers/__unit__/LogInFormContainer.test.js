import React from "react";
import { shallow } from "enzyme";
import { LogInFormContainer } from "../LogInFormContainer";

describe("LogInFormContainer", () => {
  let props;
  beforeEach(() => {
    props = {
      id: "abc",
      onSubmit: jest.fn(),
      onRegister: jest.fn(),
      onResetPassword: jest.fn(),
      onLogin: jest.fn()
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<LogInFormContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  describe("#handleUsernameChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new LogInFormContainer(props);
      subject.setState = jest.fn();
      subject.handleUsernameChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        username: "boogers",
        errorMessage: null
      });
    });
  });
  describe("#handlePasswordChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new LogInFormContainer(props);
      subject.setState = jest.fn();
      subject.handlePasswordChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        password: "boogers",
        errorMessage: null
      });
    });
  });
  describe("#handleSubmit", () => {
    it("maintains the pendingRequest state value", () => {
      expect.assertions(4);
      props.onSubmit.mockResolvedValue();
      const subject = new LogInFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false
        });
      });
      return promise;
    });
    it("sets the error message if the call fails", () => {
      expect.assertions(4);
      props.onSubmit.mockRejectedValue("fuck you");
      const subject = new LogInFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false,
          errorMessage: "fuck you"
        });
      });
      return promise;
    });
  });
});
