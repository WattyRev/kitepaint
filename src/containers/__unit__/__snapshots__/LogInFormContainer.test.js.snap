// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInFormContainer renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LogInFormContainer
    id="abc"
    onRegister={[MockFunction]}
    onResetPassword={[MockFunction]}
    onSubmit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "errorMessage": null,
      "id": "abc",
      "isDisabled": false,
      "onPasswordChange": [Function],
      "onRegister": [MockFunction],
      "onResetPassword": [MockFunction],
      "onSubmit": [Function],
      "onUsernameChange": [Function],
      "password": "",
      "username": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errorMessage": null,
        "id": "abc",
        "isDisabled": false,
        "onPasswordChange": [Function],
        "onRegister": [MockFunction],
        "onResetPassword": [MockFunction],
        "onSubmit": [Function],
        "onUsernameChange": [Function],
        "password": "",
        "username": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
